sdk/pinocchio/src/sysvars/instructions.rs

```diff
@@ -0,0 +1,287 @@
+    })
+}
+
+/// Load an `Instruction` in the currently executing `Transaction` at the
+/// specified index.
+///
+/// `data` is the instructions sysvar account data.
+///
+/// # Safety
+/// 
+/// This function is unsafe because it does not verify the address of the sysvar account.
+/// 
+/// It is typically used internally within the `load_instruction_at_checked` function, which
+/// performs the necessary address verification. However, to optimize performance for users 
+/// who have already conducted this check elsewhere in their code, we have exposed it as an 
+/// unsafe function rather than making it private, as was done in the original implementation.
+#[inline(always)]
+pub unsafe fn load_instruction_at(
+    index: usize, 
+    data: &[u8]
+) -> Result<IntrospectedInstruction, SanitizeError> {
+    deserialize_instruction(index, data)
+}
+
+/// Load an `Instruction` in the currently executing `Transaction` at the
+/// specified index.
+/// 
+/// # Errors
+/// 
+/// Returns [`ProgramError::UnsupportedSysvar`] if the given account's ID is not equal to [`ID`].
+/// Returns [`ProgramError::InvalidArgument`] if the index is out of bounds.
+#[inline(always)]
+pub fn load_instruction_at_checked(
+    index: usize,
+    instruction_sysvar_account_info: &AccountInfo,
+) -> Result<IntrospectedInstruction, ProgramError> {
+    if instruction_sysvar_account_info.key() != &INSTRUCTIONS_ID {
+        return Err(ProgramError::UnsupportedSysvar);
+    }
+
+    let instruction_sysvar_data = instruction_sysvar_account_info.try_borrow_data()?;
+    unsafe {
+        load_instruction_at(index, &instruction_sysvar_data)
+            .map(|instr| instr.clone())
+            .map_err(|err| match err {
+                SanitizeError::IndexOutOfBounds => ProgramError::InvalidArgument,
+                _ => ProgramError::InvalidInstructionData,
+            })
+    }
+}
+
+/// Returns the `Instruction` relative to the current `Instruction` in the
+/// currently executing `Transaction`.
+///
+/// # Errors
+///
+/// Returns [`ProgramError::UnsupportedSysvar`] if the given account's ID is not equal to [`ID`].
+/// Returns [`ProgramError::InvalidArgument`] if the index is out of bounds.
+pub fn get_instruction_relative(
+    index_relative_to_current: i64,
+    instruction_sysvar_account_info: &AccountInfo,
+) -> Result<IntrospectedInstruction, ProgramError> {
+    if instruction_sysvar_account_info.key() != &INSTRUCTIONS_ID {
+        return Err(ProgramError::UnsupportedSysvar);
+    }
+
+    let current_index = load_current_index_checked(&instruction_sysvar_account_info)? as i64;
+    let index = current_index.saturating_add(index_relative_to_current);
+    if index < 0 {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let instruction_sysvar_data = instruction_sysvar_account_info.try_borrow_data()?;
+    unsafe {
+        load_instruction_at(index as usize, &instruction_sysvar_data)
+            .map(|instr| instr.clone())
```

Comment:
You don't need to clone here because `load_instruction_at` returns an owned `IntrospectedInstruction` struct, yeah? Small performance improvement.